dnscache.go

```diff
@@ -0,0 +1,130 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+type DNSCache struct {
+	resolver *net.Resolver
+	mutex    sync.Mutex
+	size     int
+	duration time.Duration
+	entries  map[string]*dnsCacheEntry
+}
+
+func NewDNSCache(size int, duration time.Duration) *DNSCache {
+	return &DNSCache{
+		resolver: net.DefaultResolver,
+		size:     size,
+		duration: duration,
+		entries:  make(map[string]*dnsCacheEntry),
+	}
+}
+
+type dnsCacheEntry struct {
+	addrs   []net.IPAddr
+	expires time.Time
+}
+
+func (c *DNSCache) lookup(ctx context.Context, name string) (*dnsCacheEntry, bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
Should we be holding onto the lock during `LookupIPAddr`?

Reply:
Oh, that's a good catch, I'll fix that.